_posts/2020-09-07-sparklyr-flint/sparklyr_flint.Rmd

```diff
@@ -0,0 +1,227 @@
+```
+
+- Create a simple Spark dataframe containing some time-series data:
+
+```{scala, eval = FALSE, echo = TRUE, indent = "  "}
+import spark.implicits._
+
+val ts_sdf = Seq((1L, 1), (2L, 4), (3L, 9), (4L, 16)).toDF("time", "value")
+```
+
+- Import the dataframe along with additional metadata such as time unit and name of the timestamp column into a `TimeSeriesRDD` so that `Flint` can interpret the time-series data unambiguously:
+
+```{scala, eval = FALSE, echo = TRUE, indent = "  "}
+import com.twosigma.flint.timeseries.TimeSeriesRDD
+
+val ts_rdd = TimeSeriesRDD.fromDF(
+  ts_sdf
+)(
+  isSorted = true, // rows are already sorted by time
+  timeUnit = java.util.concurrent.TimeUnit.SECONDS,
+  timeColumn = "time"
+)
+```
+
+- Finally, after all the hard work above, we can leverage various time-series functionalities provided by `Flint` to analyze `ts_rdd`. For example, the following will produce a new column named `value_sum`. For each row, `value_sum` will contain the summation of `value`s occurred within the past 2 seconds from the timestamp of that row:
+
+```{scala, eval = FALSE, echo = TRUE, indent = "  "}
+import com.twosigma.flint.timeseries.Windows
+import com.twosigma.flint.timeseries.Summarizers
+
+val window = Windows.pastAbsoluteTime("2s")
+val summarizer = Summarizers.sum("value")
+val result = ts_rdd.summarizeWindows(window, summarizer)
+
+result.toDF.show()
+```
+
+
+```
+    +-------------------+-----+---------+
+    |               time|value|value_sum|
+    +-------------------+-----+---------+
+    |1970-01-01 00:00:01|    1|      1.0|
+    |1970-01-01 00:00:02|    4|      5.0|
+    |1970-01-01 00:00:03|    9|     14.0|
+    |1970-01-01 00:00:04|   16|     29.0|
+    +-------------------+-----+---------+
+```
+
+|      In other words, given a timestamp `t` and a row in the result having `time` equal to `t`, one can notice the `value_sum` column of that row contains sum of `value`s within the time window of `[t - 2, t]` from `ts_rdd`.
+
+# Intro to `sparklyr.flint`
+
+The purpose of `sparklyr.flint` is to make time-series functionalities of `Flint` easily accessible from `sparklyr`. To see `sparklyr.flint` in action, one can skim through the example in the previous section, go through the following to produce the exact R-equivalent of each step in that example, and then obtain the same summarization as the final result:
+
+- First of all, install `sparklyr` and `sparklyr.flint` if you haven't done so already.
+
+```{r, eval = FALSE, echo = TRUE, indent = "  "}
+install.packages("sparklyr")
+install.packages("sparklyr.flint")
+```
+
+- Connect to Apache Spark that is running locally from `sparklyr`, but remember to attach `sparklyr.flint` before running `sparklyr::spark_connect`, and then import our example time-series data to Spark:
+
+```{r, eval = FALSE, echo = TRUE, indent = "  "}
+library(sparklyr)
+library(sparklyr.flint)
+
+sc <- spark_connect(master = "local", spark_home = "~/spark/spark-2.4.4-bin-hadoop2.7")
```

Comment:
Users might not have `spark_home` set to that particular path, specially in Windows, so it's preferable to connect with the version and let `sparklyr` figure out where `spark_home` is:

```suggestion
sc <- spark_connect(master = "local", version = "2.4.4")
```

Reply:
I mean, `version = "2.4"`, not `version = "2.4.4"`, since sparklyr is smart enough to figure out the minor version.